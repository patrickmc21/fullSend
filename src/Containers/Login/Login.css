.login-page {
  height: 100vh;
  background-image: url('./images/background.jpg');
  background-size: cover;
  background-repeat: no-repeat;
  box-sizing: border-box;
  text-align: center;
}

.login-container {
  padding-top: 250px;
  margin: auto;
  height: 100%;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  box-sizing: border-box;
}

.login-logo {
  margin-bottom: 40px;
  color: #F9F8F2;
  font-family: 'Sedgwick Ave Display', cursive;
  font-size: 56px;
}

.authorize-strava {
  border: none;
  height: 40px;
  width: 160px;
  background-color: transparent;
  background-image: url('../../strava-brand-images/btn_strava_connectwith_orange.svg');
  background-size: cover;
  box-sizing: content-box;
}

.authorize-strava:hover {
  border: 2px solid #CECCC1;
}

.enter-site {
  padding: 2px 10px;
  border: 1px solid #EB592A;
  border-radius: 5px;
  height: 40px;
  width: 160px;
  background-color: #EB592A;
  color: #F9F8F2;
  font-family: 'Raleway', sans-serif;
  font-size: 42px;
  text-align: center;
  text-decoration: none;
}

.enter-site:hover {
  background-color: #F9F8F2;
  color: #EB592A;
}




