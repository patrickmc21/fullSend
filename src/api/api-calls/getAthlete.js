export const redirectLogin = () => {
  window.location = 'https://www.strava.com/oauth/authorize?client_id=24677&response_type=code&redirect_uri=http://localhost:3000&approval_prompt=force';
}