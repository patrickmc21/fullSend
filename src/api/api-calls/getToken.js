const getToken = async (temporaryToken) => {
    const history = window.location.href;
    const token = history.substr(history.length - 40);
    const url = 'http://localhost:3100/tokenexchange';
    const options = {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        mode: 'cors',
        cache: 'no-cache',
        body: JSON.stringify({
            token: token
        })
    }

    const response = await fetch(url, options)
    const data = await response.json();
    return data;
}

export default getToken