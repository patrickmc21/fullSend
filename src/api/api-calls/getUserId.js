export const getUserId = async (user) => {
  const { email, password } = user;
  const url = 'fullsend/users/signin';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      email,
      password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  };

  const response = await fetch(url, options);
  console.log(response);
  if (response.ok) {
    const id = await response.json();
    return id.id; 
  } else {
    console.log(response.error)
    return null;
  }
}