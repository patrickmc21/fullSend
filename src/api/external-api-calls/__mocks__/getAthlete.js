export const redirectLogin = jest.fn();