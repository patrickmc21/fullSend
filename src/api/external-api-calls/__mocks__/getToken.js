const getToken = jest.fn()
  .mockImplementation((token) => {
    if (token === 1) {
      return {
        access_token: 1,
        athlete: {
          name: 'Bob',
          email: 'cool@msn.com',
          password: 123
        }
      }
    } else {
      return {
        access_token: 2,
        athlete: {
          name: 'Time',
          email: 'lame@aol.com',
          password: 321
        }
      }
    }
  });

export default getToken;