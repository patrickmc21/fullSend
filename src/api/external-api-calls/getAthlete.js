export const redirectLogin = () => {
  const url = 'https://www.strava.com/oauth/authorize';
  const urlOptions = '?client_id=24677&response_type=code&';
  const redirect = 'redirect_uri=http://http://fullsend.surge.sh/&approval_prompt=force';
  window.location = `${url}${urlOptions}${redirect}`;
};