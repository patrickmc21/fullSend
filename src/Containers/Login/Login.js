import React, { Component } from 'react';
import { connect } from 'react-redux';
import { NavLink, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import { redirectLogin } from '../../api/external-api-calls/getAthlete';
import getToken from '../../api/external-api-calls/getToken';
import getUserId from '../../api/internal-api-calls/getUserId';
import createUserId from '../../api/internal-api-calls/createUserId';
import getRides from '../../api/internal-api-calls/getUserRides';
import * as actions from '../../Actions';
import './Login.css';

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      redirected: false,
      tempToken: '',
      errorStatus: ''
    };
  }

  componentDidMount() {
    this.handleRedirection(window.location);
  }

  handleRedirection = (location) => {
    if (location.search.includes('code')) {
      const history = location.href;
      const tempToken = history.substr(history.length - 40);
      this.setState(
        {
          redirected: true,
          tempToken
        }
      );
    }
  };

  handleClickAuthorize = () => {
    redirectLogin();
  };

  handleClickEnter = async () => {
    const userId = await this.loginUser();
    this.getUserRides(userId);
  };

  loginUser = async () => {
    const athleteInfo = await getToken(this.state.tempToken);
    /* eslint-disable camelcase */
    const { access_token, athlete } = athleteInfo;
    /* eslint-enable camelcase */
    const signInInfo = { email: athlete.email, password: athlete.id};
    let userId = await getUserId(signInInfo);
    if (!userId) {
      const newUser = {name: athlete.firstname, ...signInInfo};
      try {
        userId = await createUserId(newUser);
      } catch (error) {
        this.setState({errorStatus: error.message});
      }
    }
    const user = {
      name: athlete.firstname,
      /* eslint-disable camelcase */
      token: access_token,
      /* eslint-enable camelcase */
      id: userId ? userId.id : null,
      stravaId: athlete.id
    };
    this.props.addUser(user);
    return userId ? userId.id : null;
  };

  getUserRides = async (userId) => {
    const userRides = await getRides(userId);
    this.props.updateRides(userRides);
  }



  render() {
    const { redirected, errorStatus } = this.state;
    return (
      <div className='login-page'>
        <div className='login-container'>
          <h1 className='login-logo'>fullSend</h1>
          {!redirected && 
            <button 
              className='authorize-strava' 
              onClick={this.handleClickAuthorize}>
            </button>}
          {redirected && 
            <div className='enter-container'>
              <NavLink 
                to='/main' 
                onClick={this.handleClickEnter}
                className='enter-site'>
                  Enter
              </NavLink>
            </div>
          }
          { errorStatus && <p className='error-status'>{errorStatus}</p>}
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  addUser: PropTypes.func,
  updateRides: PropTypes.func
};

export const mapDispatchToProps = dispatch => ({
  addUser: user => dispatch(actions.signInUser(user)),
  updateRides: rides => dispatch(actions.updateRides(rides))
});

export default withRouter(connect(null, mapDispatchToProps)(Login));
