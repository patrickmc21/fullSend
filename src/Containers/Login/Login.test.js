import React from 'react';
import { shallow } from 'enzyme';
import { redirectLogin } from '../../api/external-api-calls/getAthlete';
import getToken from '../../api/external-api-calls/getToken';
import { getUserId } from '../../api/internal-api-calls/getUserId';
import { createUserId } from '../../api/internal-api-calls/createUserId';
import getRides from '../../api/internal-api-calls/getUserRides';

jest.mock('../../api/external-api-calls/getAthlete');
jest.mock('../../api/external-api-calls/getToken');
jest.mock('../../api/internal-api-calls/getUserId');
jest.mock('../../api/internal-api-calls/createUserId');
jest.mock('../../api/internal-api-calls/getUserRides');

import { Login, mapDispatchToProps } from './Login';

describe('Login', () => {

  let wrapper;
  let mockAddUser;
  let mockAddRides;
  let mockToken;
  let mockLocation;

  beforeEach(() => {
    mockToken = '12345qwert09876jhgfd75849quthf76453bfhgj';
    mockLocation = {
      search: 'main',
      href: mockToken
    }
    mockAddUser = jest.fn();
    mockAddRides = jest.fn();
    wrapper = shallow(
      <Login
        addUser={mockAddUser}
        addRides={mockAddRides} />
    );
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot once redirected', () => {
    wrapper.setState({redirected: true});
    expect(wrapper).toMatchSnapshot();
  });

  it('should run handleRedirection on mount', () => {
    wrapper = shallow(
      <Login
        addUser={mockAddUser}
        addRides={mockAddRides} />,
      {disableLifecycleMethods: true}
    );
    const spy = jest.spyOn(wrapper.instance(), 'handleRedirection');
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  it('should set redirected state if location includes code', () => {
    wrapper = shallow(
      <Login
        addUser={mockAddUser}
        addRides={mockAddRides} />,
      {disableLifecycleMethods: true}
    );
    mockLocation.search = 'code';
    const expected = true;
    wrapper.instance().handleRedirection(mockLocation);
    expect(wrapper.state().redirected).toEqual(expected);
  });

  it('should set tempToken state if location includes code', () => {
    wrapper = shallow(
      <Login
        addUser={mockAddUser}
        addRides={mockAddRides} />,
      {disableLifecycleMethods: true}
    );
    mockLocation.search = 'code';
    const expected = mockToken;
    wrapper.instance().handleRedirection(mockLocation);
    expect(wrapper.state().tempToken).toEqual(expected);
  });

  it('should run redirectLogin on handleClickAuthorize', () => {
    wrapper.instance().handleClickAuthorize();
    expect(redirectLogin).toHaveBeenCalled();
  });

  it('should run getToken on running loginUser', async () => {
    wrapper.setState({tempToken: 1});
    const expected = wrapper.state().tempToken;
    await wrapper.instance().loginUser();
    expect(getToken).toHaveBeenCalledWith(expected);
  });  

});

describe('mapDispatchToProps', () => {

});