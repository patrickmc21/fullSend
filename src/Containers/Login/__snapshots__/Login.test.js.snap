// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    addRides={[Function]}
    addUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="login-container"
>
        <h1
                className="login-logo"
        >
                fullSend
        </h1>
        <button
                className="authorize-strava"
                onClick={[Function]}
        />
</div>,
      "className": "login-page",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="login-logo"
>
            fullSend
</h1>,
          <button
            className="authorize-strava"
            onClick={[Function]}
/>,
          false,
          undefined,
        ],
        "className": "login-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "fullSend",
            "className": "login-logo",
          },
          "ref": null,
          "rendered": "fullSend",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "authorize-strava",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
        false,
        undefined,
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="login-container"
>
          <h1
                    className="login-logo"
          >
                    fullSend
          </h1>
          <button
                    className="authorize-strava"
                    onClick={[Function]}
          />
</div>,
        "className": "login-page",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="login-logo"
>
              fullSend
</h1>,
            <button
              className="authorize-strava"
              onClick={[Function]}
/>,
            false,
            undefined,
          ],
          "className": "login-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "fullSend",
              "className": "login-logo",
            },
            "ref": null,
            "rendered": "fullSend",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "authorize-strava",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
          false,
          undefined,
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Login should match the snapshot once redirected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    addRides={[Function]}
    addUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="login-container"
>
        <h1
                className="login-logo"
        >
                fullSend
        </h1>
        <div
                className="enter-container"
        >
                <NavLink
                        activeClassName="active"
                        ariaCurrent="true"
                        className="enter-site"
                        onClick={[Function]}
                        to="/main"
                >
                        Send It
                </NavLink>
        </div>
</div>,
      "className": "login-page",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="login-logo"
>
            fullSend
</h1>,
          false,
          <div
            className="enter-container"
>
            <NavLink
                        activeClassName="active"
                        ariaCurrent="true"
                        className="enter-site"
                        onClick={[Function]}
                        to="/main"
            >
                        Send It
            </NavLink>
</div>,
          undefined,
        ],
        "className": "login-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "fullSend",
            "className": "login-logo",
          },
          "ref": null,
          "rendered": "fullSend",
          "type": "h1",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <NavLink
              activeClassName="active"
              ariaCurrent="true"
              className="enter-site"
              onClick={[Function]}
              to="/main"
>
              Send It
</NavLink>,
            "className": "enter-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activeClassName": "active",
              "ariaCurrent": "true",
              "children": "Send It",
              "className": "enter-site",
              "onClick": [Function],
              "to": "/main",
            },
            "ref": null,
            "rendered": "Send It",
            "type": [Function],
          },
          "type": "div",
        },
        undefined,
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="login-container"
>
          <h1
                    className="login-logo"
          >
                    fullSend
          </h1>
          <div
                    className="enter-container"
          >
                    <NavLink
                              activeClassName="active"
                              ariaCurrent="true"
                              className="enter-site"
                              onClick={[Function]}
                              to="/main"
                    >
                              Send It
                    </NavLink>
          </div>
</div>,
        "className": "login-page",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="login-logo"
>
              fullSend
</h1>,
            false,
            <div
              className="enter-container"
>
              <NavLink
                            activeClassName="active"
                            ariaCurrent="true"
                            className="enter-site"
                            onClick={[Function]}
                            to="/main"
              >
                            Send It
              </NavLink>
</div>,
            undefined,
          ],
          "className": "login-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "fullSend",
              "className": "login-logo",
            },
            "ref": null,
            "rendered": "fullSend",
            "type": "h1",
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <NavLink
                activeClassName="active"
                ariaCurrent="true"
                className="enter-site"
                onClick={[Function]}
                to="/main"
>
                Send It
</NavLink>,
              "className": "enter-container",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "active",
                "ariaCurrent": "true",
                "children": "Send It",
                "className": "enter-site",
                "onClick": [Function],
                "to": "/main",
              },
              "ref": null,
              "rendered": "Send It",
              "type": [Function],
            },
            "type": "div",
          },
          undefined,
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
